src/drivers/dw/dma.c

```diff
@@ -1079,12 +1088,17 @@ static int dw_dma_free_data_size(struct dma_chan_data *channel)
 	int32_t write_ptr = dw_chan->ptr_data.current_ptr;
 	int size;
 
+	dw_chan->ptr_data.hw_ptr = read_ptr;
+
 	size = read_ptr - write_ptr;
-	if (size < 0)
+	if (size < 0) {
 		size += dw_chan->ptr_data.buffer_bytes;
-
-	if (!size)
-		tr_info(&dwdma_tr, "dw_dma_free_data_size() size is 0!");
+	} else if (!size) {
+		if (dw_chan->ptr_data.limit)
```

Comment:
and here, so that anyone reading will understand the correct buffer calculation flow (since this was missed in the past).